### 문제 1.

가야할 거리와 현재 가진 돈을 입력 받아 택시를 탈 수 있는지 여부를 출력하는 프로그램을 작성하라.

* input() 함수를 두 번 활용해 가야할 거리와 현재 가진 돈을 입력 받아 각각 distance, money 변수에 할당한다.
* 가야할 거리, 즉 distance의 단위는 km이다.
* 택시 요금은 기본 요금과 거리 비례 요금의 합으로 계산된다.
* 기본 요금은 3,000원이며 거리비례 요금은 2km 초과한 거리에 대하여 1km당 704원으로 계산된다.
* 시간에 따른 추가 요금은 없으며 계산된 택시 요금은 10원단위로 절하된다. (예. 계산된 금액이 3,709.95원인 경우 택시 요금은 3,700원)

<b>채점 기준</b>
* 결과 값이 천 단위 컴마로 구분되지 않은 경우 -1점
* 결과 값이 10원 단위로 절하되지 않은 경우 -1점
* 음수 입력, 문자열 입력 등 모든 종류의 오류를 발생시킬 수 있는 입력에 대하여 처리되지 않은 경우 -1점

*문제에 명확하게 언급되지 않았더라도 실행 예시를 참고해 실행 예시와 똑같은 결과가 나오도록 코드를 작성한다.

<i>참고.</i> 천 단위 컴마로 구분하는 방법
<pre><code>
won = 30000
print ("{:,d}".format(won))
>>> 30,000
</code></pre>

In [1]:
#while문을 이용하여 제대로 입력되었을 경우에는 멈추지 않고 계속 값을 입력할 수 있도록 해놓았습니다.
#문자가 입력되었거나, 음수가 입력되었을 경우에는 자동으로 while문을 빠져나오게 됩니다.

while True :
    try :
        distance = input("가야할 거리를 입력하세요 : ")
        money = input("현재 가진 돈을 입력하세요 : ")

        distance = float(distance)
        money = int(money)

        if distance < 0 or money < 0:
            print("입력이 제대로 되지 않았습니다.")
            break


        if distance <= 2 :
            fare = 3000
        else :
            fare = 3000 + (distance-2)*704

        fare = int(fare / 10)*10
        print ("택시 요금은 {:,d}원입니다.".format(fare))

        if money < fare :
            print("따라서 택시를 탈 수 없습니다.\n")
        else :
            print("따라서 택시를 탈 수 있습니다.\n")


    except ValueError: #character가 입력될 경우
        print("입력이 제대로 되지 않았습니다.")
        break

가야할 거리를 입력하세요 : 5
현재 가진 돈을 입력하세요 : 10000
택시 요금은 5,110원입니다.
따라서 택시를 탈 수 있습니다.

가야할 거리를 입력하세요 : 2
현재 가진 돈을 입력하세요 : 2500
택시 요금은 3,000원입니다.
따라서 택시를 탈 수 없습니다.

가야할 거리를 입력하세요 : -2
현재 가진 돈을 입력하세요 : 3000
입력이 제대로 되지 않았습니다.


실행 예 1. 모든 입력 정상 - 택시탈 수 있을 때
<pre><code>
가야할 거리를 입력하세요 : 5
현재 가진 돈을 입력하세요 : 10000
택시 요금은 5,110원입니다.
따라서 택시를 탈 수 있습니다.
</code></pre>
실행 예 2. 모든 입력 정상 - 택시탈 수 없을 때
<pre><code>
가야할 거리를 입력하세요 : 2
현재 가진 돈을 입력하세요 : 2500
택시 요금은 3,000원입니다.
따라서 택시를 탈 수 없습니다.
</code></pre>
실행 예 3. 입력 오류 - 논리적 오류
<pre><code>
가야할 거리를 입력하세요 : -2
현재 가진 돈을 입력하세요 : 3000
입력이 제대로 되지 않았습니다.
</code></pre>
실행 예 4. 입력 오류 - 런타임 오류
<pre><code>
가야할 거리를 입력하세요 : a90
현재 가진 돈을 입력하세요 : 10000
입력이 제대로 되지 않았습니다.
</code></pre>

### 문제 2.

시간(분) m을 입력받아 현재시간으로부터 m분 후의 시간을 출력하는 프로그램을 작성하라.

* input() 함수를 한 번 활용해 분 단위 시간을 입력받아 변수 m에 할당한다.
* 모든 입력은 1이상의 자연수라고 가정한다.
* 현재 시간으로부터 m분 후의 시간을 계산해 "XX분 뒤 YY시 ZZ분입니다."라고 출력한다.
* 시간의 표시는 24시간제를 사용한다. 따라서 시간은 0이상 23이하의 값을 갖는다.

*문제에 명확하게 언급되지 않았더라도 실행 예시를 참고해 실행 예시와 똑같은 결과가 나오도록 코드를 작성한다. (예. 현재시간표시)

<i>참고.</i> 현재 시스템 상 시간을 받아 변수에 저장하는 방법
<pre><code>
from datetime import datetime
now = datetime.now()
hour = now.hour
minute = now.minute

print ("현재 {}시 {}분입니다.".format(hour, minute))
>>> 현재 20시 59분입니다.
</code></pre>

**현재 시간을 받아오는 부분 이외에 별도의 모듈, 패키지, 라이브러리 사용은 허용되지 않는다.**

In [2]:
from datetime import datetime

In [3]:
m = int(input("몇 분 뒤의 시간을 알고 싶은신가요 : "))
now = datetime.now()
hour = now.hour #현재 시간의 시
minute = now.minute #현재 시간의 분

print ("현재 {}시 {}분입니다.".format(hour, minute))
hour = hour + (minute + m) // 60
minute = (minute + m) % 60
print ("{}분 뒤 {}시 {}분입니다.".format(m, hour, minute))

몇 분 뒤의 시간을 알고 싶은신가요 : 27
현재 19시 55분입니다.
27분 뒤 20시 22분입니다.


실행 예 1.
<pre><code>
몇 분 뒤의 시간을 알고 싶은신가요 : 1
현재 20시 59분입니다.
1분 뒤 21시 0분입니다.
</code></pre>
실행 예 2.
<pre><code>
몇 분 뒤의 시간을 알고 싶은신가요 : 131
현재 20시 59분입니다.
131분 뒤 23시 10분입니다.
</code></pre>
실행 예 3.
<pre><code>
몇 분 뒤의 시간을 알고 싶은신가요 : 181
현재 20시 59분입니다.
181분 뒤 0시 0분입니다.
</code></pre>
실행 예 4.
<pre><code>
몇 분 뒤의 시간을 알고 싶은신가요 : 0
현재 20시 59분입니다.
0분 뒤 20시 59분입니다.
</code></pre>
실행 예 5.
<pre><code>
몇 분 뒤의 시간을 알고 싶은신가요 : 240
현재 20시 59분입니다.
240분 뒤 0시 59분입니다.
</code></pre>

### 문제 3.

임의의 한국어 문장을 입력 받아 문장의 주어에 해당하는 어절의 목록을 출력하는 함수 get_subjects_kor()를 정의하자.

* 입력 받을 전달인자는 임의의 문장이며 문자열 자료형이다.
* 처리 결과는 문장에서 주어에 해당하는 어절들을 담은 리스트 자료형으로 반환한다.
* 각 어절은 문장 안에서 띄어쓰기로 구분되고 띄어쓰기는 항상 옳다고 가정한다.
* 주어 판정은 주격 조사를 기준으로 한다. 주격 조사는 "은", "는", "이", "가"로 한정한다.
* 따라서 어느 정도의 오류는 허용한다.
* 예. "날고 있는 나비는 아름답다."와 같은 문장에서 "있는"은 주어가 아니지만 주어로 판정된다.

*문제에 명확하게 언급되지 않았더라도 실행 예시를 참고해 실행 예시와 똑같은 결과가 나오도록 코드를 작성한다.

<i>참고.</i> 문자열이 어떤 문자열로 끝나는지 확인하는 메쏘드 str.endswith()
<pre><code>
s = "Butter Fly Effect"
if s.endswith("e"):
    print ("e로 끝납니다.")
elif s.endswith("ct"):
    print ("ct로 끝납니다.")
else:
    print ("e나 ct로 끝나지 않습니다.")
>>> ct로 끝납니다.
</code></pre>

In [4]:
# 수정 금지
test_set = ["르노 클리오는 해치백이다.",
            "세금 중 개소세는 개별소비세의 줄임말이다.",
            "르노 브랜드가 탄생한 국가는 프랑스이다.",
            "연비를 향상시키는 운전 방법은 정속 주행이다.",
            "모든 차의 운전자는 철길 건널목 통과 전에는 일시정지해야한다.",
            "올해 기준 1종 운전면허 필기시험 커트라인 점수는 80점이다.",
            "르노를 상징하는 엠블럼의 명칭은 로장주이다.",
            "일반 차량의 국내 고속도로 최고 제한속도는 경찰청 고시 기준 110km/h이다.",
            "루이 르노의 설립으로 시작된 르노의 역사는 올해로 120주년이다.",
            "엔진 브레이크는 자동차의 기어를 이용해서 속도를 줄이는 방법이다.",
            "르노 클리오의 전 세대의 글로벌 판매량은 1400만대 이상이다.",
            "올해 기준 일반 승용차 자동차 연비 등급 중 1등급은 복합연비 16km/L이상이다.",
            "신차 안전도 평가 EURO NCAP에서 르노 클리오 4세대가 받은 평가등급은 5 Stars이다.",
            "눈길에서는 타이어 공기압을 정상 수준보다 10~15% 높게 유지해야 좋다.",
            "고장난 신차를 교환이나 환불 받을 수 있는 법이 내년 도입될 예정이다.",
            "레몬법이 바로 그 법이다.",
            "르노 클리오는 유럽 올해의 차를 2회 수상하였다.",
            "세계 최초로 자동차 번호판 사용을 의무화한 프랑스",
            "메간은 국내에 정식 출시된 르노 차량이 아니다.",
            "택배차량을 분류하는 번호판 문자가 있다.",
            "최고 연비를 자랑하는 르노 클리오의 연비는 복합연비 기준 17.7km/L이다.",
            "인제 양양터널",
            "와이퍼 블레이드 교환 서비스",
            "2018년 7월 기준 우리나라 자동차 등록 대수는 약 2,290만대"]
answer_set = [['클리오는'], ['개소세는'], ['브랜드가', '국가는'], ['향상시키는', '방법은'],
              ['운전자는', '전에는'], ['점수는'], ['상징하는', '명칭은'], ['제한속도는'],
              ['루이', '역사는'], ['브레이크는', '줄이는'], ['판매량은'], ['1등급은'],
              ['평가', '4세대가', '받은', '평가등급은'], ['눈길에서는'], ['있는', '법이'],
              ['레몬법이'], ['클리오는'], [], ['메간은', '차량이'], ['분류하는', '문자가'],
              ['자랑하는', '연비는'], [], [], ['대수는']]

In [7]:
postposition = ["은", "는", "이", "가"] #조사 목록

In [8]:
sent = test_set[0].split() 
for word in sent :
    for post in postposition :
        if word.endswith(post) :
            print(word)

클리오는


위의 코드를 list comprehension 형식으로 바꾼 것이 아래 코드입니다.

In [9]:
[ word for word in sent for post in postposition if word.endswith(post) ]

['클리오는']

In [10]:
def get_subjects_kor(sent):
    sent = sent.split() #한 문장을 띄어쓰기단위로 나눈다
    postposition = ["은", "는", "이", "가"] #조사 목록
    
    return [ word for word in sent for post in postposition if word.endswith(post) ]

get_subjects_kor() 함수가 제대로 정의된 경우 아래 코드를 수행했을 때 정확도가 100.00%가 나와야한다.

In [12]:
# 수정 금지
result = {True:0, False:0}
for _, sent in enumerate(test_set):
    result[get_subjects_kor(sent) == answer_set[_]] += 1

print ("정확도 : {:.2f}%".format(100 * result[True] / (result[True] + result[False])))

정확도 : 100.00%
